td-shim-tools/src/bin/td-payload-reference-calculator/main.rs

```diff
@@ -31,6 +43,95 @@ fn param(param: &str, size: &str) -> Result<String> {
     padding_digest(param, siz)
 }
 
+fn qemu(path: &str, size: &str) -> Result<String> {
+    let path = Path::new(path).to_path_buf();
+    let siz = parse::<u64>(size)?;
+    let file_size = std::fs::metadata(&path)?.len();
+    if file_size > siz {
+        bail!("File size should be less than `kernel-size`");
+    }
+    let buf = std::fs::read(path)?;
+    qemu_patch(buf)
+}
+
+fn qemu_patch(mut buf: Vec<u8>) -> Result<String> {
+    // refer to https://github.com/qemu/qemu/blob/f48c205fb42be48e2e47b7e1cd9a2802e5ca17b0/hw/i386/x86.c#L999
+    // patching type_of_loader @0x210
+    buf[0x210] = 0xb0;
+
+    // refer to https://github.com/qemu/qemu/blob/f48c205fb42be48e2e47b7e1cd9a2802e5ca17b0/hw/i386/x86.c#L1003
+    // patching loadflags @0x211
+    buf[0x211] = 0x81;
+
+    // refer to https://github.com/qemu/qemu/blob/9c74490bff6c8886a922008d0c9ce6cae70dd17e/hw/i386/x86.c#L1004
+    // patching heap_end_ptr @0x224 cmdline_addr - real_addr - 0x200 = 0xfe00
+    buf[0x224] = 0x00;
+    buf[0x225] = 0xfe;
+
+    // refer to https://github.com/qemu/qemu/blob/9c74490bff6c8886a922008d0c9ce6cae70dd17e/hw/i386/x86.c#L962
+    // patching cmd_line_ptr @0x228 cmdline_addr = 0x20000
+    buf[0x228] = 0x00;
+    buf[0x229] = 0x00;
+    buf[0x22A] = 0x02;
+
+    let mut hasher = sha2::Sha384::new();
+    let (number_of_region_entry, regions_base, regions_size) = get_image_regions(&buf);
+
+    for index in 0..number_of_region_entry {
+        hasher.update(&buf[regions_base[index]..regions_base[index] + regions_size[index]]);
+    }
+
+    let res = hasher.finalize();
+    Ok(hex::encode(res))
+}
+
+fn get_image_regions(buf: &[u8]) -> (usize, Vec<usize>, Vec<usize>) {
+    // These 3 regions are known.
+    let mut number_of_region_entry = 3;
+    let mut regions_base = vec![
+        IMGAE_BEGIN_CHECKSUM_ADDR,
+        IMGAE_CERT_TABLE_ADDR,
+        IMGAE_HEADERS_ADDR,
+    ];
+    let mut regions_size = vec![
+        IMGAE_BEGIN_CHECKSUM_SIZE,
+        IMGAE_CERT_TABLE_SIZE,
+        IMGAE_HEADERS_SIZE,
+    ];
+
+    // Refer to https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#coff-file-header-object-and-image,
+    // After the signature of an image file is COFF File Header
+    // the NumberOfSections' offset is 2
+    let coff_file_header_offset = ((buf[IMAGE_PE_OFFSET + 3] as u32) << 24)
+        | ((buf[IMAGE_PE_OFFSET + 2] as u32) << 16)
+        | ((buf[IMAGE_PE_OFFSET + 1] as u32) << 8)
+        | (buf[IMAGE_PE_OFFSET] as u32) + PE_SIGNATURE_SIZE;
+    let number_of_pecoff_entry = buf[coff_file_header_offset as usize + 2];
```

Comment:
The spec indicates this field is 2 bytes. I'm referring to `NumberOfSections` [here](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#coff-file-header-object-and-image). That said, it also reads: "Note that the Windows loader limits the number of sections to 96." So, maybe you're fine.

Reply:
Yes. It's better to get 2 bytes here. Have corrected.